### PR TITLE
Feature/perf regression tests

### LINKED ISSUE #117: [Feature] Automated tests for performance regression
**Feature (What you would like to be added):**
We should add automated tests to check for performance regression at least before every release if not during every build.

**Motivation (Why is this needed?):**
Catch issues like #102 and #115 as early as possibe.

**Approach/Hint to the implement solution (optional):**
With the memory leak fixed in #116, the memory usage is predictable again.

A simple way to test regression could be as follows.
1. Deploy `etcd` and the backup-restore sidecar with predefined [resource limits](https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/#resource-requests-and-limits-of-pod-and-container) based on the expected database size during the test.
1. Run`etcd` [benchmark](https://github.com/etcd-io/etcd/tree/master/tools/benchmark) (multiple times if necessary) to simulate load to create desired database size.
1. Verify that the `etcd` and `backup-restore` sidecars are still healthy at the end of the test and are never restarted due to the memory limits.

### PR DESCRIPTION
**Pull request by amshuman-kr**
**What this PR does / why we need it**:
Introduces the `make perf-regression-test` target to run performance regression tests.
The tests can be executed manually against a Kubernetes cluster pointed to by the environment variable `PERF_TEST_KUBECONFIG`. This would help in early identification of regression in performance  due to changes in the future if run before every release.

**Which issue(s) this PR fixes**:
Fixes #117. Automatic execution of performance regression in CI/CD is still pending.

**Special notes for your reviewer**:
Integration with [test machinery](https://github.com/gardener/test-infra/) is not done yet, but  the tests and scripts are written in a way that can easily integrate with it.

**Release note**:

```improvement operator
Performance regression tests can be executed using the `make perf-regression-test` target against the Kubernetes cluster pointed to by the environment variable PERF_TEST_KUBECONFIG.
```

### CODE REVIEW

File: test/perf/regression/target.go [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+	for _, resourcePath := range t.resourcePaths {
+		t.logger.Infof("Creating resource from %s", resourcePath)
+
+		var (
+			u   *unstructured.Unstructured
+			gvr *schema.GroupVersionResource
+			ri  dynamic.ResourceInterface
+			err error
+		)
+		{
```
Comment (swapnilgm) [inline test/perf/regression/target.go]: Why this scope?

File: test/perf/regression/target.go [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+				}
+			}
+
+			u = &unstructured.Unstructured{}
+			err = u.UnmarshalJSON(renderedJSON)
+			if err != nil {
+				return err
+			}
+		}
+		{
```
Comment (swapnilgm) [inline test/perf/regression/target.go]: Why this scope?

File: test/perf/regression/target.go [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+		return false, err
+	}
+
+	for i := range podList.Items {
+		p := &podList.Items[i]
+		if p.Status.Phase != corev1.PodRunning {
+			return false, nil
+		}
+
+		nNotReady := len(p.Status.ContainerStatuses)
```
Comment (swapnilgm) [inline test/perf/regression/target.go]:
```suggestion
		notReady := len(p.Status.ContainerStatuses)
```

File: test/perf/regression/target.go [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+		for _, cs := range p.Status.ContainerStatuses {
+			if cs.State.Terminated != nil {
+				return false, fmt.Errorf("Container %s the namespace %s terminated unexpectedly", cs.Name, t.namespace)
+			} else if !cs.Ready {
+				return false, fmt.Errorf("Container %s in the namespace %s is not ready", cs.Name, t.namespace)
+			} else {
+				nNotReady--
```
Comment (swapnilgm) [inline test/perf/regression/target.go]:
```suggestion
				notReady--
```

File: test/perf/regression/backup_test.go [hunk 5]
```diff
@@ -0,0 +1,103 @@
+package regression
```
Comment (swapnilgm) [inline test/perf/regression/backup_test.go]: Add license header.

File: test/perf/regression/regression_suite_test.go [hunk 6]
```diff
@@ -0,0 +1,13 @@
+package regression
```
Comment (swapnilgm) [inline test/perf/regression/regression_suite_test.go]: Add license header.

File: test/perf/regression/target.go [hunk 7]
```diff
@@ -0,0 +1,265 @@
+package regression
```
Comment (swapnilgm) [inline test/perf/regression/target.go]: Add license header.

File: test/perf/regression/utils.go [hunk 8]
```diff
@@ -0,0 +1,28 @@
+package regression
```
Comment (swapnilgm) [inline test/perf/regression/utils.go]: Add license header.

File: test/perf/regression/target.go (hunk 3 above)
Comment (amshuman-kr) [reply test/perf/regression/target.go]: This is aesthetics. I would rather keep `nNotReady`, if that is alright.

File: test/perf/regression/target.go (hunk 4 above)
Comment (amshuman-kr) [reply test/perf/regression/target.go]: See above.

File: test/perf/regression/target.go (hunk 1 above)
Comment (amshuman-kr) [reply test/perf/regression/target.go]: To release `renderedJSON` for GC as soon as possibly.

File: test/perf/regression/target.go (hunk 2 above)
Comment (amshuman-kr) [reply test/perf/regression/target.go]: To restrict `gvr` and `nri` to the minimum required scope.

**Comment by amshuman-kr** [comment]
@swapnilgm I have addressed your comments and rebased. Could you PTAL?

File: test/perf/regression/resources/etcd/pod.yaml [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+    - mountPath: /var/etcd/data
+      name: etcd
+    - mountPath: /bootstrap
+      name: bootstrap
+  - name: backup-restore
+    image: {{ .etcdbrImage }}
+    command:
+    - etcdbrctl
+    - server
+    - --schedule=*/3 * * * *
```
[the remark this answers is not in the document]
Comment (amshuman-kr) [reply test/perf/regression/resources/etcd/pod.yaml]: Oops. Nice catch. The delta was meant to be `30s`. Not `300s`. Will correct and document it.

File: test/perf/regression/resources/etcd/pvc.yaml [hunk 10]
```diff
@@ -0,0 +1,11 @@
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: etcd
+  namespace: {{ .namespace }}
+spec:
+  accessModes:
```
[the remark this answers is not in the document]
Comment (amshuman-kr) [reply test/perf/regression/resources/etcd/pvc.yaml]: We can. Just wanted to keep the dependency for the test setup minimal. WDYT?

**Comment by amshuman-kr** [comment]
@swapnilgm I have [addressed](https://github.com/gardener/etcd-backup-restore/pull/171/commits/8f6a36bdb59e3322e10ce5cf9500f790e92c68a8) your review comments. Could you PTAL?

File: test/perf/regression/resources/etcd/pvc.yaml (hunk 10 above)
[the remark this answers is not in the document]
Comment (swapnilgm) [reply test/perf/regression/resources/etcd/pvc.yaml]: Fine let it be.

**Pull request merged**